File: app/src/main/java/com/zacharee1/systemuituner/fragmenthelpers/LockHelper.java
package com.zacharee1.systemuituner.fragmenthelpers;

import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.preference.Preference;
import android.preference.PreferenceCategory;
import android.preference.SwitchPreference;
import android.provider.Settings;

import com.zacharee1.systemuituner.R;
import com.zacharee1.systemuituner.activites.apppickers.AppsListActivity;
import com.zacharee1.systemuituner.fragments.ItemDetailFragment;
import com.zacharee1.systemuituner.util.SettingsUtils;

public class LockHelper extends BaseHelper {
    public static final String LOCKSCREEN_SHORTCUTS = "lockscreen_shortcuts";
    public static final String OREO_NEEDED = "oreo_needed";
    public static final String KEYGUARD_LEFT_UNLOCK = "sysui_keyguard_left_unlock";
    public static final String KEYGUARD_RIGHT_UNLOCK = "sysui_keyguard_right_unlock";
    public static final String KEYGUARD_LEFT = "sysui_keyguard_left";
    public static final String KEYGUARD_RIGHT = "sysui_keyguard_right";
    public static final String CHOOSE_LEFT = "choose_left";
    public static final String CHOOSE_RIGHT = "choose_right";
    public static final String EXTRA_ISLEFT = "isLeft";
    public static final String RESET_LEFT = "reset_left";
    public static final String RESET_RIGHT = "reset_right";

    public LockHelper(ItemDetailFragment fragment) {
        super(fragment);
        setEnabled();
        setLockIconStuff();
        setShortcutSwitchListeners();
        setResetListeners();
    }

    @Override
    public void onDestroy() {

    }

    @Override
    public void onResume() {
        setLockSummaryAndIcon();
    }

    private void setEnabled() {
        PreferenceCategory shortcuts = (PreferenceCategory) findPreference(LOCKSCREEN_SHORTCUTS);
        Preference oreoMsg = findPreference(OREO_NEEDED);
        boolean isOreo = Build.VERSION.SDK_INT > Build.VERSION_CODES.N_MR1;

        shortcuts.setEnabled(isOreo);
        if (isOreo) shortcuts.removePreference(oreoMsg);
    }

    private void setShortcutSwitchListeners() {
        SwitchPreference left = (SwitchPreference) findPreference(KEYGUARD_LEFT_UNLOCK);
        SwitchPreference right = (SwitchPreference) findPreference(KEYGUARD_RIGHT_UNLOCK);
        Preference.OnPreferenceChangeListener    listener = new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                SettingsUtils.writeSecure(getContext(), preference.getKey(), newValue.toString());
                return true;
            }
        };

        left.setOnPreferenceChangeListener(listener);
        right.setOnPreferenceChangeListener(listener);
    }

    private void setLockIconStuff() {
        setLockSummaryAndIcon();

        final Preference left = findPreference(CHOOSE_LEFT);
        left.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent activity = new Intent(getContext(), AppsListActivity.class);
                activity.putExtra(EXTRA_ISLEFT, true);
                startActivityForResult(activity, 1337);
                return true;
            }
        });

        Preference right = findPreference(CHOOSE_RIGHT);
        right.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent activity = new Intent(getContext(), AppsListActivity.class);
                activity.putExtra(EXTRA_ISLEFT, false);
                startActivityForResult(activity, 1337);
                return true;
            }
        });
    }

    private void setLockSummaryAndIcon() {
        Preference leftLock = findPreference(CHOOSE_LEFT);
        Preference rightLock = findPreference(CHOOSE_RIGHT);

        String leftSum = Settings.Secure.getString(getContext().getContentResolver(), KEYGUARD_LEFT);
        String rightSum = Settings.Secure.getString(getContext().getContentResolver(), KEYGUARD_RIGHT);

        String[] leftStuff = null;
        String[] rightStuff = null;

        if (leftSum != null) leftStuff = leftSum.split("[/]");
        if (rightSum != null) rightStuff = rightSum.split("[/]");

//        if (leftSum != null) leftSum = leftSum.replace("/.", "");
//        if (rightSum != null) rightSum = rightSum.replace("/.", "");

        leftLock.setSummary(leftSum);
        rightLock.setSummary(rightSum);

        PackageManager pm = getActivity().getPackageManager();

        Drawable unknown = getContext().getResources().getDrawable(R.drawable.ic_help_outline_black_24dp, null);
        unknown.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP);

        try {
            leftLock.setIcon(pm.getActivityIcon(new ComponentName(leftStuff[0], leftStuff[1])));
        } catch (Exception e) {
            leftLock.setIcon(unknown);
        }

        try {
            rightLock.setIcon(pm.getActivityIcon(new ComponentName(rightStuff[0], rightStuff[1])));
        } catch (Exception e) {
            rightLock.setIcon(unknown);
        }
    }

    private void setResetListeners() {
        Preference resetLeft = findPreference(RESET_LEFT);
        Preference resetRight = findPreference(RESET_RIGHT);

        resetLeft.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                SettingsUtils.writeSecure(getContext(), KEYGUARD_LEFT, "");
                setLockSummaryAndIcon();
                return true;
            }
        });
        resetRight.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                SettingsUtils.writeSecure(getContext(), KEYGUARD_RIGHT, "");
                setLockSummaryAndIcon();
                return true;
            }
        });
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/fragmenthelpers/LockHelper.kt
package com.zacharee1.systemuituner.fragmenthelpers

import android.content.ComponentName
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Build
import android.preference.Preference
import android.preference.PreferenceCategory
import android.preference.SwitchPreference
import android.provider.Settings
import com.zacharee1.systemuituner.R
import com.zacharee1.systemuituner.activites.apppickers.AppsListActivity
import com.zacharee1.systemuituner.fragments.ItemDetailFragment
import com.zacharee1.systemuituner.util.SettingsUtils

class LockHelper(fragment: ItemDetailFragment) : BaseHelper(fragment) {

    init {
        setEnabled()
        setLockIconStuff()
        setShortcutSwitchListeners()
        setResetListeners()
    }

    override fun onDestroy() {

    }

    override fun onResume() {
        setLockSummaryAndIcon()
    }

    private fun setEnabled() {
        val shortcuts = findPreference(LOCKSCREEN_SHORTCUTS) as PreferenceCategory
        val oreoMsg = findPreference(OREO_NEEDED)
        val isOreo = Build.VERSION.SDK_INT > Build.VERSION_CODES.N_MR1

        shortcuts.isEnabled = isOreo
        if (isOreo) shortcuts.removePreference(oreoMsg)
    }

    private fun setShortcutSwitchListeners() {
        val left = findPreference(KEYGUARD_LEFT_UNLOCK) as SwitchPreference
        val right = findPreference(KEYGUARD_RIGHT_UNLOCK) as SwitchPreference
        val listener = Preference.OnPreferenceChangeListener { preference, newValue ->
            SettingsUtils.writeSecure(context, preference.key, newValue.toString())
            true
        }

        left.onPreferenceChangeListener = listener
        right.onPreferenceChangeListener = listener
    }

    private fun setLockIconStuff() {
        setLockSummaryAndIcon()

        val left = findPreference(CHOOSE_LEFT)
        left?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val activity = Intent(context, AppsListActivity::class.java)
            activity.putExtra(EXTRA_ISLEFT, true)
            startActivityForResult(activity, 1337)
            true
        }

        val right = findPreference(CHOOSE_RIGHT)
        right?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val activity = Intent(context, AppsListActivity::class.java)
            activity.putExtra(EXTRA_ISLEFT, false)
            startActivityForResult(activity, 1337)
            true
        }
    }

    private fun setLockSummaryAndIcon() {
        val leftLock = findPreference(CHOOSE_LEFT)
        val rightLock = findPreference(CHOOSE_RIGHT)

        val leftSum = Settings.Secure.getString(context.contentResolver, KEYGUARD_LEFT)
        val rightSum = Settings.Secure.getString(context.contentResolver, KEYGUARD_RIGHT)

        var leftStuff: Array<String>? = null
        var rightStuff: Array<String>? = null

        if (leftSum != null) leftStuff = leftSum.split("/").toTypedArray()
        if (rightSum != null) rightStuff = rightSum.split("/").toTypedArray()

        leftLock?.summary = leftSum
        rightLock?.summary = rightSum

        val pm = activity?.packageManager

        val unknown = context.resources.getDrawable(R.drawable.ic_help_outline_black_24dp, null)
        unknown.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP)

        try {
            leftLock?.icon = pm?.getActivityIcon(ComponentName(leftStuff!![0], leftStuff[1]))
        } catch (e: Exception) {
            leftLock?.icon = unknown
        }

        try {
            rightLock?.icon = pm?.getActivityIcon(ComponentName(rightStuff!![0], rightStuff[1]))
        } catch (e: Exception) {
            rightLock?.icon = unknown
        }

    }

    private fun setResetListeners() {
        val resetLeft = findPreference(RESET_LEFT)
        val resetRight = findPreference(RESET_RIGHT)

        resetLeft?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            SettingsUtils.writeSecure(context, KEYGUARD_LEFT, "")
            setLockSummaryAndIcon()
            true
        }
        resetRight?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            SettingsUtils.writeSecure(context, KEYGUARD_RIGHT, "")
            setLockSummaryAndIcon()
            true
        }
    }

    companion object {
        const val LOCKSCREEN_SHORTCUTS = "lockscreen_shortcuts"
        const val OREO_NEEDED = "oreo_needed"
        const val KEYGUARD_LEFT_UNLOCK = "sysui_keyguard_left_unlock"
        const val KEYGUARD_RIGHT_UNLOCK = "sysui_keyguard_right_unlock"
        const val KEYGUARD_LEFT = "sysui_keyguard_left"
        const val KEYGUARD_RIGHT = "sysui_keyguard_right"
        const val CHOOSE_LEFT = "choose_left"
        const val CHOOSE_RIGHT = "choose_right"
        const val EXTRA_ISLEFT = "isLeft"
        const val RESET_LEFT = "reset_left"
        const val RESET_RIGHT = "reset_right"
    }
}
